7486 reduce tooltip over group

LucioChavezFuentes opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

My changes consists of:

1. Setting max boundaries with `maxWidth` and `numberOfLines` props.
2. Get the text width based on how it is wrapped on max boundaries.
3. Set the blackbox width according to text width on `componentDidMount`.
4. Move `measureTooltip` function, `tooltipWidth` and `tooltipHeight` states from `Tooltip/index.js`  to `TooltipRenderedOnPageBody` component.

### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.


Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/7486

### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
1. Create a new group on chats side bar.

![image](https://user-images.githubusercontent.com/48106652/161828778-bf9e36cb-991e-425b-b58c-8715a607b84f.png)

2. Add enough users to the group so the amount of emails' texts could make a large text.

3. Hover over MultipleAvatars icon of your recently created group on Chats side bar.

![image](https://user-images.githubusercontent.com/48106652/161657938-577807f6-a4df-4340-aca1-a42380cca668.png)

4. A multilined tooltip not wider than `variables.sideBarWidth` should appear with no extra white space on right side of emails.

![image](https://user-images.githubusercontent.com/48106652/161658107-86d59571-c194-4f56-be1b-a8e834c4268f.png)

5. If the emails are long enough to fit more than 3 lines then no more lines should show and elipsis (...) appear at the end of the text.

![image](https://user-images.githubusercontent.com/48106652/161658644-43286ef0-1a9d-40e1-8de4-3a50a04334b5.png)

6. Also check this setting on the following:

  a. Chat title end elipsis (...).

![image](https://user-images.githubusercontent.com/48106652/162545830-98e00e4d-587f-4f23-9645-265543eb04bb.png)

  b. On Report Header title end elipsis
![image](https://user-images.githubusercontent.com/48106652/162545779-a6abf06f-73c3-48e3-98c8-6b6a788e5eef.png)

  c. chat title end elipsis on Search Page.
![image](https://user-images.githubusercontent.com/48106652/162545900-52b68506-0002-4f8d-a497-dee27df306cd.png)


- [x] Verify that no errors appear in the JS console



### PR Review Checklist
<!--
This is a checklist for PR authors & reviewers. Please make sure to complete all tasks and check them off once you do, or else Expensify has the right not to merge your PR!
-->
#### Contributor (PR Author) Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I ran the tests on **all platforms** & verified they passed on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose and it is
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn’t already exist
    - [x] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.


<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [] I verified the correct issue is linked in the `### Fixed Issues` section above
- [] I verified testing steps are clear and they cover the changes made in this PR
    - [] I verified the steps for local testing are in the `Tests` section
    - [] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
9. Upload an image via copy paste
10. Verify a modal appears displaying a preview of that image
--->

The steps are the same of **Tests** section.

- [x] Verify that no errors appear in the JS console

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->
![image](https://user-images.githubusercontent.com/48106652/161829762-ebeb9156-9b64-4324-be08-002d2690c081.png)


#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->
**N/A**
#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->
![image](https://user-images.githubusercontent.com/48106652/161830445-25258efe-10fa-4a6d-8dbc-8b25de15d30e.png)

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

**N/A**
#### Android
<!-- Insert screenshots of your changes on the Android platform-->

**N/A**

LucioChavezFuentes committed:
convert to class, get ref, add styles

LucioChavezFuentes committed:
finish proposal

LucioChavezFuentes committed:
move getCorrectWidth func

LucioChavezFuentes committed:
create tooltipHorizontalPadding, add new param

LucioChavezFuentes committed:
remove unneeded maximumWords prop

LucioChavezFuentes committed:
simplify code

LucioChavezFuentes committed:
remove unneeded maximumWords prop

LucioChavezFuentes committed:
insert maxWidth and numberLines props, move measureTooltip

LucioChavezFuentes committed:
Merge branch 'main' into 7486_reduce-tooltip-over-group

LucioChavezFuentes committed:
remove maxWidth numberOfLines default values

LucioChavezFuentes committed:
remove displayName

LucioChavezFuentes committed:
fix minor capital letter comments

Signed-off-by: LucioChavezFuentes <luciobertinchavez@gmail.com>

LucioChavezFuentes committed:
set tooltipTextWidth and maxWidth as optional

Signed-off-by: LucioChavezFuentes <luciobertinchavez@gmail.com>

parasharrajat commented on src/components/Tooltip/TooltipRenderedOnPageBody.js:
```diff
@@ -45,49 +39,95 @@ const propTypes = {
+            tooltipHeight: 0,
+        };
+
+        this.textRef = null;
```
```suggestion

```

parasharrajat commented on src/components/Tooltip/TooltipRenderedOnPageBody.js:
```diff
@@ -45,49 +39,95 @@ const propTypes = {
+        };
+
+        this.textRef = null;
+
```
```suggestion
```

parasharrajat commented on src/components/Tooltip/TooltipRenderedOnPageBody.js:
```diff
@@ -45,49 +39,95 @@ const propTypes = {
+
 TooltipRenderedOnPageBody.propTypes = propTypes;
 TooltipRenderedOnPageBody.defaultProps = defaultProps;
-TooltipRenderedOnPageBody.displayName = 'TooltipRenderedOnPageBody';
```
Shouldn't we change the memo and use PureComponent now?

parasharrajat commented on src/components/Tooltip/index.js:
```diff
@@ -200,13 +184,12 @@ class Tooltip extends PureComponent {
                         wrapperHeight={this.state.wrapperHeight}
-                        tooltipWidth={this.state.tooltipWidth}
-                        tooltipHeight={this.state.tooltipHeight}
                         setTooltipRef={el => this.tooltip = el}
```
Do we need this now?
```suggestion
```

parasharrajat commented on src/components/Tooltip/index.js:
```diff
@@ -43,7 +40,6 @@ class Tooltip extends PureComponent {
         this.animation = new Animated.Value(0);
 
         this.getWrapperPosition = this.getWrapperPosition.bind(this);
-        this.measureTooltip = this.measureTooltip.bind(this);
```
Remove line 35 as well.

parasharrajat commented on src/components/Tooltip/tooltipPropTypes.js:
```diff
@@ -24,6 +24,13 @@ const propTypes = {
     /** Any additional amount to manually adjust the vertical position of the tooltip.
     A positive value shifts the tooltip down, and a negative value shifts it up. */
     shiftVertical: PropTypes.oneOfType([PropTypes.number, PropTypes.func]),
+
+    /** Number of pixels to set max-width on tooltip  */
+    maxWidth: PropTypes.number,
+
+    /** Maximum number of lines to set on tooltip */
+    numberOfLines: PropTypes.number,
+
```
Shouldn't we pass the default values as well?

parasharrajat replied:
```suggestion
```

parasharrajat commented on src/styles/getTooltipStyles.js:
```diff
@@ -109,6 +113,11 @@ export default function getTooltipStyles(
             ...tooltipVerticalPadding,
             ...spacing.ph2,
             zIndex: variables.tooltipzIndex,
+            maxWidth: tooltipTextWidth >= maxWidth
```
These should be optional right.

LucioChavezFuentes replied:
Yep, they are optional, if maxWidth or tooltipTextWidth are undefined then the ternary operation will return a non valid value which for CSS will mean not setting max-width at all, making tooltip work as currently does.

parasharrajat replied:
How come? I think it will return false and then a second statement will be run.

LucioChavezFuentes replied:
If `maxWidth` is undefined so will be the`tootipTextWidth`. So when the second statement runs it will try to make this operation: `undefined  + number` which will return `NaN` which is non valid value for CSS. `tootipTextWidth` won't update with recent change in componentDidMount so this operation won't be executed twice. But now I think about it, I will move this logic to a function and make early void return if prop maxWidth is not set. What do you think?

LucioChavezFuentes replied:
My function to set `maxWidth` will return `undefined` or `Number` but according to  [JSDocs style guidelines](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs) it says 'Bad' using `*` on `@returns` on example shown and can't use union types. What should I add on `@returns`?

![image](https://user-images.githubusercontent.com/48106652/162489801-ec1cef13-2178-4ed8-b219-917840e8e707.png)

parasharrajat replied:
if you notice carefully, we are defining the maxWidth of the object literal. So it should be easy to conditionally add props to the literal value. 

There are many ways to do this. 
1. we create a variable for the returning value. And then add the new props conditionally. 
2. we create a variable whose value is set conditionally to the object of maxWidth, defaults to `{}`. Then we expand this into the return value

LucioChavezFuentes committed:
avoid set state on tooltips with no maxWidth

LucioChavezFuentes committed:
remove memo, make component Pure, minor changes

LucioChavezFuentes commented on src/components/Tooltip/TooltipRenderedOnPageBody.js:
```diff
@@ -39,59 +33,102 @@ const propTypes = {
+    }
+
+    componentDidMount() {
+        if (!this.props.maxWidth) {
+            return;
```
I added a checking to avoid `setState` for tooltips with no `maxWidth`.

LucioChavezFuentes commented on src/components/Tooltip/TooltipRenderedOnPageBody.js:
```diff
@@ -39,59 +33,102 @@ const propTypes = {
+        this.measureTooltip = this.measureTooltip.bind(this);
+    }
+
+    componentDidMount() {
+        if (!this.props.maxWidth) {
+            return;
+        }
+        this.setState({
+            tooltipTextWidth: this.textRef.offsetWidth,
+        });
+    }
```
I added a checking to avoid `setState` for tooltips with no `maxWidth`.

parasharrajat commented:
Also, this tooltip settings needs to be applied a few more locations. 
1. LHN chat title end. on `...`
2. same on SearchPage. 
3. And all the other pages where we showing this tooltip. 
4. Report Header. Title of the report page.

LucioChavezFuentes committed:
tooltip settings, create setWrapperWidth

LucioChavezFuentes commented:
> Also, this tooltip settings needs to be applied a few more locations.
> 
> 1. LHN chat title end. on `...`
> 2. same on SearchPage.
> 3. And all the other pages where we showing this tooltip.
> 4. Report Header. Title of the report page.

Apply Tooltip settings on DisplayNames.js, it seems it cover all. I checked the app fairly well and I didn't find more tooltps that need this setting but if you find one let me know.

parasharrajat commented on src/styles/getTooltipStyles.js:
```diff
@@ -49,6 +49,30 @@ function computeHorizontalShift(windowWidth, xOffset, componentWidth, tooltipWid
+        return ({maxWidth});
+    }
+
+    // Sum left and right wrapper padding, otherwise the wrapper will clip the text.
+    // Add 1 px safe buffer to avoid last word wrap or clip on special cases
+    // where the real text width is float but a little larger than ref.offsetWidth, which will return integer
+    // ex. real text width: 172.2 px, ref.offsetWidth: 172px
```
This is not clear at all. What are you trying to say here? Why does 1px need to be added?

parasharrajat commented on src/styles/getTooltipStyles.js:
```diff
@@ -49,6 +49,30 @@ function computeHorizontalShift(windowWidth, xOffset, componentWidth, tooltipWid
+ */
+function setWrapperWidth(maxWidth, tooltipTextWidth) {
+    if (!maxWidth) {
+        return ({});
```
```suggestion
        return {};
```

parasharrajat commented on src/styles/getTooltipStyles.js:
```diff
@@ -49,6 +49,30 @@ function computeHorizontalShift(windowWidth, xOffset, componentWidth, tooltipWid
+    }
+
+    if (tooltipTextWidth >= maxWidth) {
+        return ({maxWidth});
```
```suggestion
        return {maxWidth};
```

parasharrajat commented on src/styles/getTooltipStyles.js:
```diff
@@ -49,6 +49,30 @@ function computeHorizontalShift(windowWidth, xOffset, componentWidth, tooltipWid
+    // Add 1 px safe buffer to avoid last word wrap or clip on special cases
+    // where the real text width is float but a little larger than ref.offsetWidth, which will return integer
+    // ex. real text width: 172.2 px, ref.offsetWidth: 172px
+    return ({
+        maxWidth: tooltipTextWidth + (spacing.ph2.paddingHorizontal * 2) + 1,
+    });
```
```suggestion
    return {
        maxWidth: tooltipTextWidth + (spacing.ph2.paddingHorizontal * 2) + 1,
    };
```

parasharrajat commented on src/components/MultipleAvatars.js:
```diff
@@ -80,7 +81,7 @@ const MultipleAvatars = (props) => {
                         </Tooltip>
                     ) : (
-                        <Tooltip text={props.avatarTooltips.slice(1).join(', ')} absolute>
+                        <Tooltip text={props.avatarTooltips.slice(1).join(', ')} absolute numberOfLines={3} maxWidth={variables.sideBarWidth}>
```
Move `numberOfLines` to const.

LucioChavezFuentes committed:
fix comment about adding 1px, add consts, minor changes

LucioChavezFuentes commented on src/styles/getTooltipStyles.js:
```diff
@@ -49,6 +49,31 @@ function computeHorizontalShift(windowWidth, xOffset, componentWidth, tooltipWid
+    }
+
+    // Sum left and right wrapper padding, otherwise the wrapper will clip the text.
+    // Sometimes the real width of a text is a value with fractional digits (i.e. 172.2px)
+    // but the text.offsetWidth returns a truncated integer (172px) making this function
+    // to set the wrapper width slightly smaller than text but enough to make the last word
+    // of a line to wrap or clip. So is neccesary add 1px extra to the sum so wrapper contains text entirely.
```
What do you think about this sentence?

parasharrajat commented on src/components/DisplayNames/index.js:
```diff
@@ -5,6 +5,9 @@ import {propTypes, defaultProps} from './displayNamesPropTypes';
 import Text from '../Text';
+import variables from '../../styles/variables';
+
+const TOOLTIP_MAX_LINES = 3;
```
I meant in CONST.js file.

LucioChavezFuentes replied:
Oh, my bad, I wasn't aware of that file. Fixing it.

parasharrajat commented on src/styles/getTooltipStyles.js:
```diff
@@ -49,6 +49,31 @@ function computeHorizontalShift(windowWidth, xOffset, componentWidth, tooltipWid
+        return {maxWidth};
+    }
+
+    // Sum left and right wrapper padding, otherwise the wrapper will clip the text.
+    // Sometimes the real width of a text is a value with fractional digits (i.e. 172.2px)
+    // but the text.offsetWidth returns a truncated integer (172px) making this function
+    // to set the wrapper width slightly smaller than text but enough to make the last word
+    // of a line to wrap or clip. So is neccesary add 1px extra to the sum so wrapper contains text entirely.
```
```suggestion
    // Add horizontal padding to the textContainerWidth to get the wrapper width. tooltipTextWidth ignores the fractions (OffsetWidth) so add 1px to fit the text properly. 
```

How does this sound?

LucioChavezFuentes replied:
```suggestion
    // Add horizontal padding to tooltipTextWidth to get the wrapper width. 
    tooltipTextWidth ignores the fractions (OffsetWidth) so add 1px to fit the text properly. 
```
Sounds nice, I would change the word `textContainerWidth` to `tooltipTextWidth`. `textContainerWidth` sounds like another variable.

parasharrajat replied:
You can always use normal text instead of technical variables

LucioChavezFuentes committed:
move TOOLTIP_MAX_LINES to CONST

LucioChavezFuentes committed:
change comment about 1px add, minor change

parasharrajat commented:
Is this ready for review?

parasharrajat reviewed (changes requested):
There are a few more tooltips that need this style. So I am suggesting that you move these 
```js
numberOfLines={CONST.TOOLTIP_MAX_LINES}
                                maxWidth={variables.sideBarWidth}
```
as the default value to the props so it will be applied automatically to all the places and then we can explicitly change it where needed. 

What do you think?

LucioChavezFuentes commented:
> There are a few more tooltips that need this style. So I am suggesting that you move these
> 
> ```js
> numberOfLines={CONST.TOOLTIP_MAX_LINES}
>                                 maxWidth={variables.sideBarWidth}
> ```
> 
> as the default value to the props so it will be applied automatically to all the places and then we can explicitly change it where needed.
> 
> What do you think?

Sounds good! actually I thought this was the idea until I asked for default values. Making changes.

LucioChavezFuentes commented:
> Is this ready for review?

Sorry, forgot to notify.

LucioChavezFuentes committed:
set default values numberOfLines maxWidth, remove check

LucioChavezFuentes committed:
remove unneccesary props

LucioChavezFuentes committed:
return maxWidth if tooltipTextWidth is not set

LucioChavezFuentes committed:
minor comment change

LucioChavezFuentes commented:
PR updated, ready to review.

parasharrajat commented on src/components/Tooltip/TooltipRenderedOnPageBody.js:
```diff
@@ -39,59 +33,95 @@ const propTypes = {
-TooltipRenderedOnPageBody.propTypes = propTypes;
-TooltipRenderedOnPageBody.defaultProps = defaultProps;
-TooltipRenderedOnPageBody.displayName = 'TooltipRenderedOnPageBody';
+const defaultProps = {};
```
```suggestion
```

parasharrajat commented on src/components/Tooltip/TooltipRenderedOnPageBody.js:
```diff
@@ -39,59 +33,95 @@ const propTypes = {
+}
+
+TooltipRenderedOnPageBody.propTypes = propTypes;
+TooltipRenderedOnPageBody.defaultProps = defaultProps;
```
```suggestion
```

parasharrajat commented on src/components/Tooltip/index.js:
```diff
@@ -26,24 +26,17 @@ class Tooltip extends PureComponent {
             wrapperWidth: 0,
             wrapperHeight: 0,
 
```
```suggestion
```

parasharrajat commented on src/styles/getTooltipStyles.js:
```diff
@@ -49,6 +49,24 @@ function computeHorizontalShift(windowWidth, xOffset, componentWidth, tooltipWid
 }
 
+/**
+ *
```
Comment??

parasharrajat commented on src/styles/getTooltipStyles.js:
```diff
@@ -49,6 +49,24 @@ function computeHorizontalShift(windowWidth, xOffset, componentWidth, tooltipWid
+    }
+
+    // Add horizontal padding to the text width to get the wrapper width.
+    //  tooltipTextWidth ignores the fractions (OffsetWidth) so add 1px to fit the text properly.
```
```suggestion
    // TooltipTextWidth ignores the fractions (OffsetWidth) so add 1px to fit the text properly.
```

parasharrajat reviewed (changes requested):
testing now.

parasharrajat commented:
![image](https://user-images.githubusercontent.com/24370807/163768621-61be8d4f-13d5-457f-8e9d-18f47f28ac03.png)

@LucioChavezFuentes Could you please show me the screenshots for the screens where you were facing whitespace? Can't find the old ones you shared.

LucioChavezFuentes commented:
> Could you please show me the screenshots for the screens where you were facing whitespace?

Sure, here is one.

![image](https://user-images.githubusercontent.com/48106652/163865333-3a04eb90-7d2b-4c09-b0f7-21a969ce83c5.png)


> ![image](https://user-images.githubusercontent.com/24370807/163772167-357999ea-d7e0-42e6-8543-666114766540.png)
The tooltip width is very narrow on the previous screenshot. There is enough room and shouldn't it expand to the maxWidth.

Let me check, do you happen to resize the window after hovering the tooltip?

parasharrajat commented:
No. Check the tooltip on report title.

LucioChavezFuentes committed:
avoid re-render on single line tooltips, minor changes

LucioChavezFuentes commented:
> ![image](https://user-images.githubusercontent.com/24370807/163772167-357999ea-d7e0-42e6-8543-666114766540.png)
The tooltip width is very narrow on the previous screenshot. There is enough room and shouldn't it expand to the maxWidth.

I'm still investigating about this matter. It looks like it doesn't happen all the time and it fixes when you reload page. I'll post when I have something more clear.

parasharrajat commented:
Thank you. Looking forward to it.

LucioChavezFuentes committed:
use width instead of max-width

LucioChavezFuentes committed:
remove check on componentDidMount

LucioChavezFuentes commented:
> ![image](https://user-images.githubusercontent.com/24370807/163772167-357999ea-d7e0-42e6-8543-666114766540.png)
The tooltip width is very narrow on the previous screenshot. There is enough room and shouldn't it expand to the maxWidth.

Fixed, I changed to `width` CSS prop instead of `max-width`. `max-width` cause to `measureTooltip` fire multiple times and setting tooltip styles wich in turn cause to give me unaccurate `text.offsetWidth`.

@parasharrajat PR updated and ready to reviewed!

LucioChavezFuentes committed:
remove unneccesary function, simplify code

LucioChavezFuentes commented:
Made a simplification, now is truly ready to be reviewed.

parasharrajat commented:
@LucioChavezFuentes Please merge main.

parasharrajat reviewed (changes requested):
Testing started...
- [ ] Ios
- [x] Android
- [ ] Desktop 
- [x] Web

parasharrajat reviewed (changes requested):
Ok, I think it is a good time to add the Stories for Tooltips.

parasharrajat commented on src/components/Tooltip/TooltipRenderedOnPageBody.js:
```diff
@@ -39,59 +33,92 @@ const propTypes = {
+        super(props);
+        this.state = {
+            // tooltipTextWidth will update with wrapped text offsetWidth based on maxWidth prop
+            tooltipTextWidth: this.props.maxWidth,
```
Could you please explain this change to me?

parasharrajat commented on src/components/Tooltip/TooltipRenderedOnPageBody.js:
```diff
@@ -39,59 +33,92 @@ const propTypes = {
+
+    componentDidMount() {
+        this.setState({
+            tooltipTextWidth: this.textRef.offsetWidth,
```
I feel like we can get this value from the `this.state.tooltipWidth`.

tooltipTextWidth = this.state.tooltipWidth - (spacing.ph2.paddingHorizontal * 2);

parasharrajat replied:
So if there is a race condition then why aren't we using the onLayout on the Text or [measure](https://reactnative.dev/docs/0.67/direct-manipulation#measurecallback).

LucioChavezFuentes replied:
Measure returns the same width values as ref.offsetWidth , actually, if you use measure and use ‘max-width’ css prop in wrapperStyle, you can reproduce the bug easily where the blackbox tooltip won’t be taking the full max-width. The same happens with measureInWindow.

And onLayout does not work on nested <Text> components.

parasharrajat replied:
The only reason to keep pushing this is that something does not feel right. Using `offsetWidth` is not the standard way in react-native. It is vanilla Js. 

But I am fine with it as far as it is working.

LucioChavezFuentes replied:
Ok, I understand now.  The bug where the blackbox tooltip doesn't take the full max-width came form `max-width` css prop. As long as we use `width` instead of `max-width` in wrapperStyle there is no difference between using `measure` or `offsetWidth`. So I can change it to `measure` to be on sync with the react-native way.

parasharrajat commented on src/styles/getTooltipStyles.js:
```diff
@@ -110,6 +114,15 @@ export default function getTooltipStyles(
+
+                // Add horizontal padding to the text width to get the wrapper width.
+                // TooltipTextWidth ignores the fractions (OffsetWidth) so add 1px to fit the text properly.
+                ? tooltipTextWidth + (spacing.ph2.paddingHorizontal * 2) + 1
```
When text is not expanding to MaxWidth which means it is taking lesser space than the MaxWidth allowed, why are we setting the width at all. let it be automatic as it was.

LucioChavezFuentes committed:
move wrapper width logic close to its variables

LucioChavezFuentes committed:
Merge branch 'main' into 7486_reduce-tooltip-over-group

LucioChavezFuentes commented:
PR updated and ready to reviewed.

Now here comes some explanations:

#### Why not 'tooltipWidth'?

> I feel like we can get this value from the `this.state.tooltipWidth`.
> 
> tooltipTextWidth = this.state.tooltipWidth - (spacing.ph2.paddingHorizontal * 2);
 
To remove the whitespace, we are setting the wrapper width based on tooltip **innertext** width, not the width of the tooltip itself.

#### Why 'width' instead of 'max-width' ?

> When text is not expanding to MaxWidth which means it is taking lesser space than the MaxWidth allowed, why are we setting the width at all. let it be automatic as it was.

![image](https://user-images.githubusercontent.com/24370807/163772167-357999ea-d7e0-42e6-8543-666114766540.png)
 
This bug was caused because we get an inaccurate tooltip **innertext** width and that happened because `measureTooltip` was called multiple times even after `componentDidMount` (which sets the tooltip **innertext** width) is called.
 
`measureTooltip` is called by React Native's `onLayout`, this event is executed every time the style of that component (our wrapper) changes.
 `max-width` makes extra style calculations that triggers the `onLayout` event even after the component is mounted, so it becomes a race condition between `componentDidMount` and `onLayout` event. This doesn't happen all the time (it was hard for me to reproduce the bug) but it happens more commonly to tooltips close to the edge of the screen because there are extra style calculations to position them properly.
 
If we use `width`, the browser does not make extra calculations, it just sets the size of our tooltip, prevents more `onLayout` calls than needed, and then we update it with our tooltip **innertext** width.

parasharrajat reviewed (changes requested):
Thanks for the explanation. You just refactored the same thing. I will take a look the code myself. But for that we will have to wait. Thanks.

LucioChavezFuentes commented:
Sure, no problem, I'm staying tuned for any update.

LucioChavezFuentes committed:
remove unneeded wrapperView initializing

LucioChavezFuentes commented:
I just made a [minor change](https://github.com/Expensify/App/pull/8494/commits/7895f0158072f9e08ebbde93c8ade490b04a760b), forgot to remove a line.

parasharrajat commented on src/components/Tooltip/TooltipRenderedOnPageBody.js:
```diff
@@ -39,59 +34,99 @@ const propTypes = {
+    }
+
+    render() {
+        // We get wrapper width based on tooltip's inner text width so the wrapper
+        // is just big enough to fit text and prevent white space
+        const wrapperWidth = this.state.tooltipTextWidth && this.state.tooltipTextWidth < this.props.maxWidth
+
+            // Add horizontal padding to the text width to get the wrapper width.
+            // TooltipTextWidth ignores the fractions (OffsetWidth) so add 1px to fit the text properly.
+            ? this.state.tooltipTextWidth + (spacing.ph2.paddingHorizontal * 2) + 1
+            : this.props.maxWidth;
```
First of all, move this to `getTooltipStyles`.

LucioChavezFuentes replied:
Done, I also changed the params order because I noticed the last two params (`manualShiftHorizontal `and `manualShiftVertical`) are optional.

LucioChavezFuentes committed:
move wrapper width condition to getTooltipStyles

LucioChavezFuentes committed:
move maxWidth before optional parameters

LucioChavezFuentes commented:
I moved the max-width condition and explained why we can't use onLayout of the Text or measure [here](https://github.com/Expensify/App/pull/8494#discussion_r863265814).

parasharrajat commented on src/styles/getTooltipStyles.js:
```diff
@@ -93,6 +97,14 @@ export default function getTooltipStyles(
     const tooltipVerticalPadding = spacing.pv1;
     const tooltipFontSize = variables.fontSizeSmall;
 
+    // We get wrapper width based on tooltip's inner text width so the wrapper
+    // is just big enough to fit text and prevent white space
```
```suggestion
    // We get wrapper width based on the tooltip's inner text width so the wrapper is just big enough to fit text and prevent white space.
```

LucioChavezFuentes replied:
Updated.

parasharrajat reviewed (changes requested):
Testting..

parasharrajat reviewed (dismissed):
LGTM. Please do the comment changes. I missed that.

cc: @Beamanator 

:ribbon: :eyes: :ribbon:  C+ reviewed

LucioChavezFuentes committed:
update comment and change offsetWidth to measure function

LucioChavezFuentes commented:
Thanks @parasharrajat, I updated the comment and change `offsetWidth` to measure function (to be on sync with the react native way).

Beamanator commented on src/components/Tooltip/tooltipPropTypes.js:
```diff
@@ -24,6 +26,12 @@ const propTypes = {
+    /** Number of pixels to set max-width on tooltip  */
+    maxWidth: PropTypes.number,
+
+    /** Maximum number of lines to set on tooltip */
```
`numberOfLines` is the maximum number of lines the tooltip will show, right? If so I think this is a bit clearer

```suggestion
    /** Maximum number of lines to show in tooltip */
```

LucioChavezFuentes replied:
Done.

Beamanator commented on src/styles/getTooltipStyles.js:
```diff
@@ -93,6 +97,13 @@ export default function getTooltipStyles(
 
+    // We get wrapper width based on tooltip's inner text width so the wrapper is just big enough to fit text and prevent white space
+    const wrapperWidth = tooltipTextWidth && tooltipTextWidth < maxWidth
+
```
Can we get rid of this line? If the linter is mad and needs an empty line before a comment, maybe we can put the whole explanation of this calculation before line 101? I just don't really like the ternary 3 lines away from the original expression

LucioChavezFuentes replied:
Done. I moved the comments before line 101. The ternary operation is close to expression.

Beamanator commented on src/styles/getTooltipStyles.js:
```diff
@@ -93,6 +97,13 @@ export default function getTooltipStyles(
+        // Add horizontal padding to the text width to get the wrapper width.
+        // TooltipTextWidth ignores the fractions (OffsetWidth) so add 1px to fit the text properly.
+        ? tooltipTextWidth + (spacing.ph2.paddingHorizontal * 2) + 1
+        : maxWidth;
```
```suggestion
        : maxWidth;

```

LucioChavezFuentes replied:
Done.

Beamanator commented on src/components/Tooltip/TooltipRenderedOnPageBody.js:
```diff
@@ -39,59 +33,94 @@ const propTypes = {
 // On every tooltip hover, we update the position in state which will result in re-rendering.
 // We also update the state on layout changes which will be triggered often.
 // There will be n number of tooltip components in the page.
 // Its good to memorize this one.
```
I know this isn't your comment, but fixing this is helpful while we're here :D

```suggestion
// It's good to memorize this one.
```

LucioChavezFuentes replied:
Done.

Beamanator commented on src/components/Tooltip/TooltipRenderedOnPageBody.js:
```diff
@@ -39,59 +33,94 @@ const propTypes = {
-        </Animated.View>,
-        document.querySelector('body'),
-    );
+    /** Maximum number of lines to set on tooltip */
```
Same as my other comment
```suggestion
    /** Maximum number of lines to show in tooltip */
```

LucioChavezFuentes replied:
Done.

Beamanator reviewed (commented):
Just a few cosmetic / comment changes requested, overall looks fantastic & tests well! 👍

parasharrajat commented on src/components/Tooltip/TooltipRenderedOnPageBody.js:
```diff
@@ -39,59 +33,94 @@ const propTypes = {
+        this.measureTooltip = this.measureTooltip.bind(this);
+    }
+
+    componentDidMount() {
+        this.textRef.measure((x, y, width) => {
+            this.setState({
+                tooltipTextWidth: width,
+            });
+        });
+    }
```
I was not saying to use the measure call here. I was thinking that we can use measure where we need the `tooltipTextWidth`. Like in the getTooltipStyles calculations. But I understand why do you need to save the width in `componentDidMount`. Using measure here is not useful at all. The previous OffsetWidth attempt was also fine. I would say it was better as it was quick. Now, this measure call is adding the async code which will further add delay. 

---
Please feel free to ask for confirmation before making changes, if you don't understand requests. It will save our time.

LucioChavezFuentes replied:
Sorry for premature pushed changes, I understand now your point of view. I tried using `measure` inside `getTooltipStyles` but I couldn't make it work  so I am reverting to `offsetWidth` solution. (along with comments changes of course).

LucioChavezFuentes committed:
revert to offsetWidth, minor comment changes

LucioChavezFuentes commented:
PR updated and ready to be reviewed.

parasharrajat reviewed (dismissed):
LGTM.

Beamanator commented on src/components/Tooltip/TooltipRenderedOnPageBody.js:
```diff
@@ -39,59 +33,92 @@ const propTypes = {
+    /**
+     * Measure the size of the tooltip itself.
+     *
+     * @param {Object} nativeEvent
```
NAB; this could be more clear - what kind of native event? And where does it come from?

LucioChavezFuentes replied:
Sorry for delay.
It comes from `Animated.View`'s  `onLayout` prop. [`onLayout`](https://reactnative.dev/docs/view#onlayout) is prop that many react-native's core components have. It returns a event object wich has another object called 'nativeEvent' which is a type of  [LayoutEvent Object](https://reactnative.dev/docs/layoutevent). The LayoutEvent object has layout info (i.e. width, height) about the component where `onLayout` prop is being used,

https://github.com/Expensify/App/blob/89c578184ea9475e4f43531eb84e5eb42d2964f8/src/components/Tooltip/TooltipRenderedOnPageBody.js#L106-L109

Maybe we can make param comment: 

```js
* @param {Object} nativeEvent - LayoutEvent
```

Beamanator replied:
Aah thanks so much for explaining - on second thought I think it's fine as is, it is kinda self explanatory and a very small function :D

Beamanator commented on src/styles/getTooltipStyles.js:
```diff
@@ -93,6 +97,13 @@ export default function getTooltipStyles(
     const tooltipVerticalPadding = spacing.pv1;
     const tooltipFontSize = variables.fontSizeSmall;
 
+    // We get wrapper width based on tooltip's inner text width so the wrapper is just big enough to fit text and prevent white space
+    // Add horizontal padding to the text width to get the wrapper width.
+    // TooltipTextWidth ignores the fractions (OffsetWidth) so add 1px to fit the text properly.
```
How about this?

```suggestion
    // We get wrapper width based on the tooltip's inner text width so the wrapper is just big enough to fit text and prevent white space.
    // If the text width is less than the maximum available width, add horizontal padding.
    // Note: tooltipTextWidth ignores the fractions (OffsetWidth) so add 1px to fit the text properly.
```

LucioChavezFuentes replied:
Sounds good! Modifying...

LucioChavezFuentes replied:
Done.

Beamanator commented on src/components/Tooltip/TooltipRenderedOnPageBody.js:
```diff
@@ -39,59 +33,92 @@ const propTypes = {
+    constructor(props) {
+        super(props);
+        this.state = {
+            // The width of toolttip's inner text
```
```suggestion
            // The width of tooltip's inner text
```

LucioChavezFuentes replied:
Done.

Beamanator reviewed (changes requested):
Thanks for sticking with us through all these requested changes 👍 Getting super close!

LucioChavezFuentes committed:
change on comments

Beamanator reviewed (approved):
Let's get this merged, thanks for sticking with this!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

LucioChavezFuentes commented:
Thank you @parasharrajat  and @Beamanator for sticking with me. I know it's been a long thread and I really appreciate your time on many reviews on this pull request.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/2320539573) to staging by @Beamanator in version: 1.1.58-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/2333359574) to production by @AndrewGable in version: 1.1.60-3 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅
